introduction_to_amazon_algorithms/jumpstart-foundation-models/text-generation-benchmarking/benchmarking/load_test.py

```diff
@@ -0,0 +1,344 @@
+
+    def client_latency(self):
+        """The client latency for this single prediction."""
+        return (self.time_utc_end - self.time_utc_start).total_seconds() * 1e3
+
+    def input_sequence_num_words(self):
+        """The word count of the input sequence."""
+        return self._num_words(self._text_inputs())
+
+    def output_sequence(self):
+        """The output sequence with the input sequence prefix removed.
+
+        Text generation models, by default, include the input sequence in the model response.
+        """
+        data = deepcopy(self.result)
+        while not isinstance(data, str):
+            if isinstance(data, list):
+                data = data[0]
+            if isinstance(data, dict):
+                if "generated_text" in data:
+                    data = data["generated_text"]
+                elif "generated_texts" in data:
+                    data = data["generated_texts"]
+                elif "generation" in data:
+                    data = data["generation"]
+                else:
+                    raise ValueError("Output data contains dictionary without recognized keys.")
+
+        text_inputs = self._text_inputs()
+        if data.startswith(text_inputs):
+            data = data[len(text_inputs) :]
+
+        return data
+
+    def num_words(self) -> int:
+        """The word count of the output sequence."""
+        return self._num_words(self.output_sequence())
+
+    def num_tokens(self, tokenizer: PreTrainedTokenizerBase) -> int:
+        """The token count of the output sequence."""
+        return len(tokenizer.encode(self.output_sequence()))
+
+    def _text_inputs(self) -> str:
+        if "inputs" in self.payload:
+            return self.payload["inputs"]
+        elif "text_inputs" in self.payload:
+            return self.payload["text_inputs"]
+        else:
+            raise ValueError("Expected input text keys are not detected in payload.")
+
+    @staticmethod
+    def _num_words(text: str) -> int:
+        return len(text.split())
+
+
+class BatchInvocationStatistics(NamedTuple):
+    """A NamedTuple holding start and stop times for a batch of endpoint predictions."""
+
+    time_utc_start: datetime.datetime
+    time_utc_end: datetime.datetime
+    num_invocations: int
+    results: List[PredictionResult]
+
+    def _duration_seconds(self) -> float:
+        """Computes the time in seconds of the batch load test."""
+        return (self.time_utc_end - self.time_utc_start).total_seconds()
+
+    def _throughput(self, values: List[int]) -> float:
+        """Computes the number of invocation responses per second."""
+        return sum(values) / self._duration_seconds()
+
+    def _throughput_robust(self, values: List[int]) -> float:
+        """Computes the median throughput of result values in units values per second to eliminate outliers."""
+        throughput_values = [
+            sum(values[: (i + 1)]) / (result.time_utc_end - self.time_utc_start).total_seconds()
```

Comment:
sanity-check: any risk of dividing by zero here if you are rounding in seconds?

Reply:
No risk, total_seconds method returns a float.